src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::{BTreeMap, HashMap};
+use std::fs::{self};
+use std::io::Cursor;
+use std::path::{Path, PathBuf};
+use std::time::Duration;
+
+use anyhow::{bail, Context};
+use byteorder::{NetworkEndian, ReadBytesExt};
+use lazy_static::lazy_static;
+use log::error;
+use regex::Regex;
+use tokio::select;
+use tokio::sync::mpsc;
+
+use dataflow_types::Update;
+use expr::GlobalId;
+use repr::{Row, Timestamp};
+
+use crate::wal::{encode_progress, encode_update};
+
+/// Organize and maintain a compacted representation of persisted data.
+///
+/// The `Compacter` task keeps track of the persisted updates for each persisted relation and
+/// periodically compacts that representation to use space proportional to the number of
+/// distinct rows in the relation. In order to do so, the `Compacter` maintains a `Trace` for
+/// each persisted relation.
+///
+/// A `Trace` is basically a list of `Batch`s that represent a contiguous time interval, and a
+/// compaction frontier.
+///
+/// A `Batch` is a consolidated list of updates that occured between times [lower, upper)
+/// where each update is of the form `(Row, time, diff)` and each `(Row, time)` pair occurs
+/// exactly once and all diffs are nonzero.
+///
+/// Note that all `Batch`s keep their data on persistent storage. No data resides on disk
```

Comment:
This line confused me about persistent storage but not disk. It may be irrelevant (maybe delete if so?) or making a point I don't understand.

Reply:
yeah I was a little bit split brained when writing these docs because I kept defaulting to use the word "disk" but then also being like "everything needs to work with s3!"

Reply:
I reread -- this comment was wrong. meant to say "No data resides in memory"